### DISCUSSION TITLE
Is this a good practice with class state?

### DISCUSSION THREAD

**Discussion by varHarrie** [pmndrs/valtio#466] [Q&A]
Full Example: https://codesandbox.io/s/valtio-forked-z5f9it

I add a `bind` util to avoid `'set' on proxy: trap returned falsish for property 'xxx'`:

```typescript
function bind<T extends object>(instance: T): T {
  const obj = instance as any;
  const names = Object.getOwnPropertyNames(Object.getPrototypeOf(obj));

  for (const name of names) {
    const method = obj[name];
    if (name === "constructor" || typeof method !== "function") continue;
    obj[name] = (...args: unknown[]) => method.apply(instance, args);
  }

  return instance;
}

class BearStore {
  water: number = 0;

  drink() {
    this.water += 1;
  }
}

const bearStore = bind(proxy(new BearStore()));

function useBearStore() {
   return useSnapshot(bearStore);
}

export default function App() {
  const bearStore = useBearStore();

  console.log('render')

  return (
    <>
      <div>{bearStore.water}</div>
      <button onClick={() => bearStore.drink()}>drink</button>
    </>
  );
}

```

There seems to be no problem. Is there something I missed?

**Comment by dai-shi** [comment]
Yeah: https://github.com/pmndrs/valtio/wiki/How-to-organize-actions#using-class

I'm not sure why you need `bind()`.

**Reply by varHarrie** [reply]
In order to forcibly bind `this` of all methods to state instance, so that I can call actions using snapshot. Based on https://github.com/pmndrs/valtio/issues/205#issuecomment-891481107

**Reply by dai-shi** [reply]
Oh, sure. I missed that point. Thanks for the clarification.

**Reply by ckken** [reply]
v2 need bind too?

**Reply by dai-shi** [reply]
I think so. Have you tried anything?
